README.md
# joi-router

Easy, rich and fully validated [koa][] routing.

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/koa-joi-router.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-joi-router
[travis-image]: https://img.shields.io/travis/koajs/joi-router.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/joi-router
[codecov-image]: https://codecov.io/github/koajs/joi-router/coverage.svg?branch=master
[codecov-url]: https://codecov.io/github/koajs/joi-router?branch=master
[david-image]: https://img.shields.io/david/koajs/joi-router.svg?style=flat-square
[david-url]: https://david-dm.org/koajs/joi-router
[download-image]: https://img.shields.io/npm/dm/koa-joi-router.svg?style=flat-square
[download-url]: https://npmjs.org/package/koa-joi-router
[co]: https://github.com/tj/co
[koa]: http://koajs.com
[co-body]: https://github.com/visionmedia/co-body
[await-busboy]: https://github.com/aheckmann/await-busboy
[joi]: https://github.com/hapijs/joi
[@koa/router]: https://github.com/koajs/router
[generate api documentation]: https://github.com/a-s-o/koa-docs
[path-to-regexp]: https://github.com/pillarjs/path-to-regexp

#### Features:

- built in input validation using [joi][]
- built in [output validation](#validating-output) using [joi][]
- built in body parsing using [co-body][] and [await-busboy][]
- built on the great [@koa/router][]
- [exposed route definitions](#routes) for later analysis
- string path support
- [regexp-like path support](#path-regexps)
- [multiple method support](#multiple-methods-support)
- [multiple middleware support](#multiple-middleware-support)
- [continue on error support](#handling-errors)
- [router prefixing support](#prefix)
- [router level middleware support](#use)
- meta data support
- HTTP 405 and 501 support

#### Node compatibility

Node.js `>= 12` is required.

#### Example

```js
const Koa = require("koa");
const KoaJoiRouter = require("koa-joi-router");
const Joi = KoaJoiRouter.Joi;

const public = new KoaJoiRouter();

public.get("/", async (ctx) => {
  ctx.body = "hello joi-router!";
});

public.route({
  method: "post",
  path: "/signup",
  validate: {
    body: {
      name: Joi.string().max(100),
      email: Joi.string().lowercase().email(),
      password: Joi.string().max(100),
      _csrf: Joi.string().token(),
    },
    type: "form",
    output: {
      200: {
        body: {
          userId: Joi.string(),
          name: Joi.string(),
        },
      },
    },
  },
  handler: async (ctx) => {
    const user = await createUser(ctx.request.body);
    ctx.status = 201;
    ctx.body = user;
  },
});

const app = new koa();
app.use(public.middleware());
app.listen(3000);
```

## Usage

`koa-joi-router` returns a class which you use to define your routes.
The design is such that you construct multiple router instances, one for
each section of your application which you then add as koa middleware.

```js
const Koa = require("koa");
const KoaJoiRouter = require("koa-joi-router");

const pub = new KoaJoiRouter();
const admin = new KoaJoiRouter();
const auth = new KoaJoiRouter();

// add some routes ..
pub.get("/some/path", async () => {});
admin.get("/admin", async () => {});
auth.post("/auth", async () => {});

const app = new Koa();
app.use(pub.middleware());
app.use(admin.middleware());
app.use(auth.middleware());
app.listen();
```

## Module properties

### .Joi

It is **HIGHLY RECOMMENDED** you use this bundled version of Joi
to avoid bugs related to passing an object created with a different
release of Joi into the router.

```js
const Koa = require("koa");
const KoaJoiRouter = require("koa-joi-router");
const Joi = KoaJoiRouter.Joi;
```

## Router instance methods

### .route()

Adds a new route to the router. `route()` accepts an object or array of objects
describing route behavior.

```js
const KoaJoiRouter = require("koa-joi-router");
const public = new KoaJoiRouter();

public.route({
  method: "post",
  path: "/signup",
  validate: {
    header: joiObject,
    query: joiObject,
    params: joiObject,
    body: joiObject,
    maxBody: "64kb",
    output: { "400-600": { body: joiObject } },
    type: "form",
    failure: 400,
    continueOnError: false,
  },
  pre: async (ctx, next) => {
    await checkAuth(ctx);
    return next();
  },
  handler: async (ctx) => {
    await createUser(ctx.request.body);
    ctx.status = 201;
  },
  meta: { this: { is: "stored internally with the route definition" } },
});
```

or

```js
const KoaJoiRouter = require("koa-joi-router");
const public = new KoaJoiRouter();

const routes = [
  {
    method: "post",
    path: "/users",
    handler: async (ctx) => {},
  },
  {
    method: "get",
    path: "/users",
    handler: async (ctx) => {},
  },
];

public.route(routes);
```

##### .route() options

- `method`: **required** HTTP method like "get", "post", "put", etc
- `path`: **required** string
- `validate`
  - `header`: object which conforms to [Joi][] validation
  - `query`: object which conforms to [Joi][] validation
  - `params`: object which conforms to [Joi][] validation
  - `body`: object which conforms to [Joi][] validation
  - `maxBody`: max incoming body size for forms or json input
  - `failure`: HTTP response code to use when input validation fails. default `400`
  - `type`: if validating the request body, this is **required**. either `form`, `json` or `multipart`
  - `formOptions`: options for co-body form parsing when `type: 'form'`
  - `jsonOptions`: options for co-body json parsing when `type: 'json'`
  - `multipartOptions`: options for [busboy][] parsing when `type: 'multipart'`
    - [any busboy constructor option][busboy]. eg `{ limits: { files: 1 }}`
    - `autoFields`: Determines whether form fields should be auto-parsed (default: `true`). See the [await-busboy docs](https://github.com/aheckmann/await-busboy#parts--parsestream-options).
  - `output`: see [output validation](#validating-output)
  - `continueOnError`: if validation fails, this flags determines if `koa-joi-router` should [continue processing](#handling-errors) the middleware stack or stop and respond with an error immediately. useful when you want your route to handle the error response. default `false`
  - `validateOptions`: options for Joi validate. default `{}`
- `handler`: **required** async function or functions
- `pre`: async function or function, will be called before parser and validators
- `meta`: meta data about this route. `koa-joi-router` ignores this but stores it along with all other route data

### .get(), post(), put(), delete() ...etc - HTTP methods

`koa-joi-router` supports the traditional `router.get()`, `router.post()` type APIs
as well.

```js
const KoaJoiRouter = require("koa-joi-router");
const admin = new KoaJoiRouter();

// signature: router.method(path [, config], handler [, handler])

admin.put("/thing", handler);
admin.get("/thing", middleware, handler);
admin.post("/thing", config, handler);
admin.delete("/thing", config, middleware, handler);
```

### .use()

Middleware run in the order they are defined by .use()(or .get(), etc.) They are invoked sequentially, requests start at the first middleware and work their way "down" the middleware stack which matches Express 4 API.

```js
const KoaJoiRouter = require("koa-joi-router");
const users = new KoaJoiRouter();

users.get("/:id", handler);
users.use("/:id", runThisAfterHandler);
```

### .prefix()

Defines a route prefix for all defined routes. This is handy in "mounting" scenarios.

```js
const KoaJoiRouter = require("koa-joi-router");
const users = new KoaJoiRouter();

users.get("/:id", handler);
// GET /users/3 -> 404
// GET /3 -> 200

users.prefix("/users");
// GET /users/3 -> 200
// GET /3 -> 404
```

### .param()

Defines middleware for named route parameters. Useful for auto-loading or validation.

_See [@koa/router](https://github.com/koajs/router/blob/master/API.md#module_koa-router--Router+param)_

```js
const KoaJoiRouter = require("koa-joi-router");
const users = new KoaJoiRouter();

const findUser = (id) => {
  // stub
  return Promise.resolve("Cheddar");
};

users.param("user", async (id, ctx, next) => {
  const user = await findUser(id);
  if (!user) return (ctx.status = 404);
  ctx.user = user;
  await next();
});

users.get("/users/:user", (ctx) => {
  ctx.body = `Hello ${ctx.user}`;
});

// GET /users/3 -> 'Hello Cheddar'
```

### .middleware()

Generates routing middleware to be used with `koa`. If this middleware is
never added to your `koa` application, your routes will not work.

```js
const KoaJoiRouter = require("koa-joi-router");
const public = new KoaJoiRouter();

public.get("/home", homepage);

const app = koa();
app.use(public.middleware()); // wired up
app.listen();
```

## Additions to ctx.state

The route definition for the currently matched route is available
via `ctx.state.route`. This object is not the exact same route
definition object which was passed into koa-joi-router, nor is it
used internally - any changes made to this object will
not have an affect on your running application but is available
to meet your introspection needs.

```js
const KoaJoiRouter = require("koa-joi-router");
const public = new KoaJoiRouter();

public.get("/hello", async (ctx) => {
  console.log(ctx.state.route);
});
```

## Additions to ctx.request

When using the `validate.type` option, `koa-joi-router` adds a few new properties
to `ctx.request` to faciliate input validation.

### ctx.request.body

The `ctx.request.body` property will be set when either of the following
`validate.type`s are set:

- json
- form

#### json

When `validate.type` is set to `json`, the incoming data must be JSON. If it is not,
validation will fail and the response status will be set to 400 or the value of
`validate.failure` if specified. If successful, `ctx.request.body` will be set to the
parsed request input.

```js
admin.route({
  method: "post",
  path: "/blog",
  validate: { type: "json" },
  handler: async (ctx) => {
    console.log(ctx.request.body); // the incoming json as an object
  },
});
```

#### form

When `validate.type` is set to `form`, the incoming data must be form data
(x-www-form-urlencoded). If it is not, validation will fail and the response
status will be set to 400 or the value of `validate.failure` if specified.
If successful, `ctx.request.body` will be set to the parsed request input.

```js
admin.route({
  method: "post",
  path: "/blog",
  validate: { type: "form" },
  handler: async (ctx) => {
    console.log(ctx.request.body); // the incoming form as an object
  },
});
```

### ctx.request.parts

The `ctx.request.parts` property will be set when either of the following
`validate.type`s are set:

- multipart

#### multipart

When `validate.type` is set to `multipart`, the incoming data must be multipart data.
If it is not, validation will fail and the response
status will be set to 400 or the value of `validate.failure` if specified.
If successful, `ctx.request.parts` will be set to an
[await-busboy][] object.

```js
admin.route({
  method: "post",
  path: "/blog",
  validate: { type: "multipart" },
  handler: async (ctx) => {
    const parts = ctx.request.parts;
    let part;

    try {
      while ((part = await parts)) {
        // do something with the incoming part stream
        part.pipe(someOtherStream);
      }
    } catch (err) {
      // handle the error
    }

    console.log(parts.field.name); // form data
  },
});
```

## Handling non-validated input

_Note:_ if you do not specify a value for `validate.type`, the
incoming payload will not be parsed or validated. It is up to you to
parse the incoming data however you see fit.

```js
admin.route({
  method: "post",
  path: "/blog",
  validate: {},
  handler: async (ctx) => {
    console.log(ctx.request.body, ctx.request.parts); // undefined undefined
  },
});
```

## Validating output

Validating the output body and/or headers your service generates on a
per-status-code basis is supported. This comes in handy when contracts
between your API and client are strict e.g. any change in response
schema could break your downstream clients. In a very active codebase, this
feature buys you stability. If the output is invalid, an HTTP status 500
will be used.

Let's look at some examples:

### Validation of an individual status code

```js
router.route({
  method: "post",
  path: "/user",
  validate: {
    output: {
      200: {
        // individual status code
        body: {
          userId: Joi.string(),
          name: Joi.string(),
        },
      },
    },
  },
  handler: handler,
});
```

### Validation of multiple individual status codes

```js
router.route({
  method: "post",
  path: "/user",
  validate: {
    output: {
      "200,201": {
        // multiple individual status codes
        body: {
          userId: Joi.string(),
          name: Joi.string(),
        },
      },
    },
  },
  handler: handler,
});
```

### Validation of a status code range

```js
router.route({
  method: "post",
  path: "/user",
  validate: {
    output: {
      "200-299": {
        // status code range
        body: {
          userId: Joi.string(),
          name: Joi.string(),
        },
      },
    },
  },
  handler: handler,
});
```

### Validation of multiple individual status codes and ranges combined

You are free to mix and match ranges and individual status codes.

```js
router.route({
  method: "post",
  path: "/user",
  validate: {
    output: {
      "200,201,300-600": {
        // mix it up
        body: {
          userId: Joi.string(),
          name: Joi.string(),
        },
      },
    },
  },
  handler: handler,
});
```

### Validation of output headers

Validating your output headers is also supported via the `headers` property:

```js
router.route({
  method: "post",
  path: "/user",
  validate: {
    output: {
      "200,201": {
        body: {
          userId: Joi.string(),
          name: Joi.string(),
        },
        headers: Joi.object({
          // validate headers too
          authorization: Joi.string().required(),
        }).options({
          allowUnknown: true,
        }),
      },
      "500-600": {
        body: {
          // this rule only runs when a status 500 - 600 is used
          error_code: Joi.number(),
          error_msg: Joi.string(),
        },
      },
    },
  },
  handler: handler,
});
```

## Router instance properties

### .routes

Each router exposes it's route definitions through it's `routes` property.
This is helpful when you'd like to introspect the previous definitions and
take action e.g. to [generate API documentation][] etc.

```js
const KoaJoiRouter = require("koa-joi-router");
const admin = new KoaJoiRouter();

admin.post("/thing", { validate: { type: "multipart" } }, handler);

console.log(admin.routes);
// [ { path: '/thing',
//     method: [ 'post' ],
//     handler: [ [Function] ],
//     validate: { type: 'multipart' } } ]
```

## Path RegExps

Sometimes you need `RegExp`-like syntax support for your route definitions.
Because [path-to-regexp][]
supports it, so do we!

```js
const KoaJoiRouter = require("koa-joi-router");
const admin = new KoaJoiRouter();

admin.get(
  "/blog/:year(\\d{4})-:day(\\d{2})-:article(\\d{3})",
  async (ctx, next) => {
    console.log(ctx.request.params); // { year: '2017', day: '01', article: '011' }
  }
);
```

## Multiple methods support

Defining a route for multiple HTTP methods in a single shot is supported.

```js
const KoaJoiRouter = require("koa-joi-router");
const admin = new KoaJoiRouter();

admin.route({
  path: "/",
  method: ["POST", "PUT"],
  handler: fn,
});
```

## Multiple middleware support

Often times you may need to add additional, route specific middleware to a
single route.

```js
const KoaJoiRouter = require("koa-joi-router");
const admin = new KoaJoiRouter();

admin.route({
  path: "/",
  method: ["POST", "PUT"],
  handler: [yourMiddleware, yourHandler],
});
```

## Nested middleware support

You may want to bundle and nest middleware in different ways for reuse and
organization purposes.

```js
const KoaJoiRouter = require("koa-joi-router");
const admin = new KoaJoiRouter();

const commonMiddleware = [yourMiddleware, someOtherMiddleware];
admin.route({
  path: "/",
  method: ["POST", "PUT"],
  handler: [commonMiddleware, yourHandler],
});
```

This also works with the .get(),post(),put(),delete(), etc HTTP method helpers.

```js
const KoaJoiRouter = require("koa-joi-router");
const admin = new KoaJoiRouter();

const commonMiddleware = [yourMiddleware, someOtherMiddleware];
admin.get("/", commonMiddleware, yourHandler);
```

## Handling errors

By default, `koa-joi-router` stops processing the middleware stack when either
input validation fails. This means your route will not be reached. If
this isn't what you want, for example, if you're writing a web app which needs
to respond with custom html describing the errors, set the `validate.continueOnError`
flag to true. You can find out if validation failed by checking `ctx.invalid`.

```js
admin.route({
  method: "post",
  path: "/add",
  validate: {
    type: "form",
    body: {
      id: Joi.string().length(10),
    },
    continueOnError: true,
  },
  handler: async (ctx) => {
    if (ctx.invalid) {
      console.log(ctx.invalid.header);
      console.log(ctx.invalid.query);
      console.log(ctx.invalid.params);
      console.log(ctx.invalid.body);
      console.log(ctx.invalid.type);
    }

    ctx.body = await render("add", { errors: ctx.invalid });
  },
});
```

## Development

### Running tests

- `npm test` runs tests + code coverage + lint
- `npm run lint` runs lint only
- `npm run lint-fix` runs lint and attempts to fix syntax issues
- `npm run test-cov` runs tests + test coverage
- `npm run open-cov` opens test coverage results in your browser
- `npm run test-only` runs tests only

## LICENSE

[MIT](https://github.com/koajs/joi-router/blob/master/LICENSE)

[busboy]: https://github.com/mscdex/busboy#busboy-methods


History.md
8.0.0 / 2021-04-26
==================

- https://github.com/koajs/joi-router/compare/9fa16b6b1..0aa5c45de5
- major bump of @koa/router from 8 to 10 - https://github.com/koajs/router/pull/77/files
  - impact: `router.get('*', ....)` becomes `router.get('(.*)') ....)`

7.0.0 / 2019-12-07
==================

- https://github.com/koajs/joi-router/compare/91f0a42e24c4..9fa16b6b1
- major bump @hapi/joi@15 -> joi@17
  - impact: changes validation error messaging. example: https://github.com/koajs/joi-router/commit/fad66d8acbb51a5ddeb5b961b0d5d3cb9646bf65

6.0.1 / 2019-12-07
==================

- bump deps

6.0.0 / 2019-04-27
==================

  * deps; [semver-major] bump @hapi/joi from 14.x -> 15.x
  * deps; bump eslint, supertest and switch from istanbul to nyc
  * lint; lint all JS #65, switch from pebble to eslint:recommended, remove eslint-plugin-standard
  * refactored; makeBodyParser #86 by swarthy

5.3.0 / 2019-03-04
==================

  * deps; bump await-busboy, koa-router, clone, co-body, debug
  * docs; update description of .use() API by wdanxna
  * tests; fix head response body check

5.2.0 / 2019-01-12
==================

 * added; pre-handler support #66 by swarthy 
 * added; allow form and json parsing options to be configured #75 by nojacko
 * fixed; always fill ctx.invalid even when continueOnError is false #70 by alvarowolfx
 * fixed; do not parse the body if already present #76 by oprogramador
 * deps; update joi to 14.0.6 #77 by pke
 * docs; parsing options #75 by nojacko

5.1.0 / 2018-01-02
==================

 * added; router.param() - #45 by pixeldrew
 * updated; to co-body 5.1.1
 * updated; to koa-router 7.3.0
 * updated; to debug 2.6.9
 * docs; clarified - #42 by paul42
 * docs; clarified - #40 by nicodinh

5.0.0 / 2017-03-09
==================

 * breaking; now requires node >= 7.6
 * added; async/await support
 * removed; generator support
 * removed; makefile in favor of npm scripts
 * deps; use await-busboy
 * deps; updated
 * docs; updated

4.0.0 / 2016-07-28
==================

 * fixed; params casting #26
 * changed; retain casted query values #25
 * removed; support for node < 4.x
 * use aheckmann/koa-router fork with params bugfix #24
 * updated; to Joi 9.0.4
 * updated; to co-body to 4.2.0
 * updated; to co-busboy to 1.3.1
 * updated; to delegates to 1.0.0
 * updated; to flatten to 1.0.2
 * updated; to methods to 1.1.2
 * docs; add koa-docs link

3.1.1 / 2016-04-27
==================

 * fixed; support node 0.12 & friends

3.1.0 / 2016-04-27
==================

 * added; route definition introspection via ctx.state.route #16 [BrainsoftLtd](https://github.com/BrainsoftLtd)
 * added; nested middleware support #14 [reyawn](https://github.com/reyawn)
 * docs; fix typo #13 [simplyianm](https://github.com/simplyianm)

3.0.0 / 2015-12-09
==================

 * BREAKING; support custom output validation per resp status code
 * added; .route() now also supports an array of routes: #7 (martinmicunda)

2.1.2 / 2015-10-10
==================

 * updated; koa-router: 5.2.3

2.1.1 / 2015-09-08
==================

 * play nice with old versions of npm

2.1.0 / 2015-09-08
==================

 * changed; use aheckmann/koa-router (until koa-router merges upstream fixes https://github.com/alexmingoia/koa-router/pull/169)
 * updated; and clean up dependencies
 * added .use() and .prefix() support
 * tests; better exercise of koa-router variations
 * tests; run tests before lint
 * add mocha env to eslint config #3 from wilmoore/eslint-mocha-env
 * add eslint

2.0.0 / 2015-08-18
==================

 * bump koa-router to 5.1.2 (RexExp are no longer first class citizens)
 * bump debug to 2.2.0
 * bump co-body to 4.0.0
 * bump joi to 6.6.1
 * bump busboy, sliced and methods deps
 * update docs
 * expose Joi module
 * support node 0.12 and iojs

1.3.1 / 2015-02-07
==================

 * fixed header validation documentation
 * refactor
 * use yield* next

1.3.0 / 2014-09-20
==================

 * rename proceed option -> continueOnError

1.2.0 / 2014-09-20
==================

 * added; support for continueOnError
 * updated; dependencies

1.1.1 / 2014-08-17
==================

 * updated docs
 * tests; for RegExp support

1.1.0 / 2014-08-15
==================

 * added; output validation support

1.0.0 / 2014-08-06
==================

 * first release


router.js
'use strict';

const assert = require('assert');
const debug = require('debug')('koa-joi-router');
const isGenFn = require('is-gen-fn');
const flatten = require('flatten');
const methods = require('methods');
const KoaRouter = require('@koa/router');
const busboy = require('await-busboy');
const parse = require('co-body');
const slice = require('sliced');
const delegate = require('delegates');
const clone = require('clone');
const OutputValidator = require('./output-validator');

module.exports = Router;

function identity(obj) {
  return obj;
}

function Router({
  validatorBuilder
} = { validatorBuilder: identity }) {
  if (!(this instanceof Router)) {
    return new Router({
      validatorBuilder
    });
  }

  this.routes = [];
  this.router = new KoaRouter();
  this.validatorBuilder = validatorBuilder
}

/**
 * Array of routes
 *
 * Router.prototype.routes;
 * @api public
 */

/**
 * Delegate methods to internal router object
 */

delegate(Router.prototype, 'router')
  .method('prefix')
  .method('use')
  .method('param');

/**
 * Return koa middleware
 * @return {Function}
 * @api public
 */

Router.prototype.middleware = function middleware() {
  return this.router.routes();
};

/**
 * Adds a route or array of routes to this router, storing the route
 * in `this.routes`.
 *
 * Example:
 *
 *   var admin = router();
 *
 *   admin.route({
 *     method: 'get',
 *     path: '/do/stuff/:id',
 *     handler: function *(next){},
 *     validate: {
 *       header: Joi object
 *       params: Joi object (:id)
 *       query: Joi object (validate key/val pairs in the querystring)
 *       body: Joi object (the request payload body) (json or form)
 *       maxBody: '64kb' // (json, x-www-form-urlencoded only - not stream size)
 *                       // optional
 *       type: 'json|form|multipart' (required when body is specified)
 *       failure: 400 // http error code to use
 *     },
 *     meta: { // this is ignored but useful for doc generators etc
 *       desc: 'We can use this for docs generation.'
 *       produces: ['application/json']
 *       model: {} // response object definition
 *     }
 *   })
 *
 * @param {Object} spec
 * @return {Router} self
 * @api public
 */

Router.prototype.route = function route(spec) {
  if (Array.isArray(spec)) {
    for (let i = 0; i < spec.length; i++) {
      this._addRoute(spec[i]);
    }
  } else {
    this._addRoute(spec);
  }

  return this;
};

/**
 * Adds a route to this router, storing the route
 * in `this.routes`.
 *
 * @param {Object} spec
 * @api private
 */

Router.prototype._addRoute = function addRoute(spec) {
  this._validateRouteSpec(spec);
  this.routes.push(spec);

  debug('add %s "%s"', spec.method, spec.path);

  const bodyParser = makeBodyParser(spec);
  const specExposer = makeSpecExposer(spec);
  const validator = makeValidator(spec, this.validatorBuilder);
  const preHandlers = spec.pre ? flatten(spec.pre) : [];
  const handlers = flatten(spec.handler);

  const args = [
    spec.path
  ].concat(preHandlers, [
    prepareRequest,
    specExposer,
    bodyParser,
    validator
  ], handlers);

  const router = this.router;

  spec.method.forEach((method) => {
    router[method].apply(router, args);
  });
};

/**
 * Validate the spec passed to route()
 *
 * @param {Object} spec
 * @api private
 */

Router.prototype._validateRouteSpec = function validateRouteSpec(spec) {
  assert(spec, 'missing spec');

  const ok = typeof spec.path === 'string' || spec.path instanceof RegExp;
  assert(ok, 'invalid route path');

  checkHandler(spec);
  checkPreHandler(spec);
  checkMethods(spec);
  checkValidators(spec);
};

/**
 * @api private
 */

function checkHandler(spec) {
  if (!Array.isArray(spec.handler)) {
    spec.handler = [spec.handler];
  }

  return flatten(spec.handler).forEach(isSupportedFunction);
}

/**
 * @api private
 */

function checkPreHandler(spec) {
  if (!spec.pre) {
    return;
  }

  if (!Array.isArray(spec.pre)) {
    spec.pre = [spec.pre];
  }

  return flatten(spec.pre).forEach(isSupportedFunction);
}

/**
 * @api private
 */

function isSupportedFunction(handler) {
  assert.equal('function', typeof handler, 'route handler must be a function');

  if (isGenFn(handler)) {
    throw new Error(`route handlers must not be GeneratorFunctions
       Please use "async function" or "function".`);
  }
}

/**
 * Validate the spec.method
 *
 * @param {Object} spec
 * @api private
 */

function checkMethods(spec) {
  assert(spec.method, 'missing route methods');

  if (typeof spec.method === 'string') {
    spec.method = spec.method.split(' ');
  }

  if (!Array.isArray(spec.method)) {
    throw new TypeError('route methods must be an array or string');
  }

  if (spec.method.length === 0) {
    throw new Error('missing route method');
  }

  spec.method.forEach((method, i) => {
    assert(typeof method === 'string', 'route method must be a string');
    spec.method[i] = method.toLowerCase();
  });
}

/**
 * Validate the spec.validators
 *
 * @param {Object} spec
 * @api private
 */

function checkValidators(spec) {
  if (!spec.validate) return;

  let text;
  if (spec.validate.body) {
    text = 'validate.type must be declared when using validate.body';
    assert(/json|form/.test(spec.validate.type), text);
  }

  if (spec.validate.type) {
    text = 'validate.type must be either json, form, multipart or stream';
    assert(/json|form|multipart|stream/i.test(spec.validate.type), text);
  }

  if (spec.validate.output) {
    spec.validate._outputValidator = new OutputValidator(spec.validate.output);
  }

  // default HTTP status code for failures
  if (!spec.validate.failure) {
    spec.validate.failure = 400;
  }
}

/**
 * Does nothing
 * @param  {[type]}   ctx  [description]
 * @param  {Function} next [description]
 * @return {async function}        [description]
 * @api private
 */

async function noopMiddleware(ctx, next) {
  return await next();
}

/**
 * Handles parser internal errors
 * @param  {Object} spec         [description]
 * @param  {function} parsePayload [description]
 * @return {async function}              [description]
 * @api private
 */

function wrapError(spec, parsePayload) {
  return async function errorHandler(ctx, next) {
    try {
      await parsePayload(ctx, next);
    } catch (err) {
      captureError(ctx, 'type', err);
      if (spec.validate.continueOnError) {
        return await next();
      } else {
        return ctx.throw(err);
      }
    }
  };
}

/**
 * Creates JSON body parser middleware.
 *
 * @param {Object} spec
 * @return {async function}
 * @api private
 */

function makeJSONBodyParser(spec) {
  const opts = spec.validate.jsonOptions || {};
  if (typeof opts.limit === 'undefined') {
    opts.limit = spec.validate.maxBody;
  }

  return async function parseJSONPayload(ctx, next) {
    if (!ctx.request.is('json')) {
      return ctx.throw(400, 'expected json');
    }

    // eslint-disable-next-line require-atomic-updates
    ctx.request.body = ctx.request.body || await parse.json(ctx, opts);
    await next();
  };
}

/**
 * Creates form body parser middleware.
 *
 * @param {Object} spec
 * @return {async function}
 * @api private
 */

function makeFormBodyParser(spec) {
  const opts = spec.validate.formOptions || {};
  if (typeof opts.limit === 'undefined') {
    opts.limit = spec.validate.maxBody;
  }
  return async function parseFormBody(ctx, next) {
    if (!ctx.request.is('urlencoded')) {
      return ctx.throw(400, 'expected x-www-form-urlencoded');
    }

    // eslint-disable-next-line require-atomic-updates
    ctx.request.body = ctx.request.body || await parse.form(ctx, opts);
    await next();
  };
}

/**
 * Creates stream/multipart-form body parser middleware.
 *
 * @param {Object} spec
 * @return {async function}
 * @api private
 */

function makeMultipartParser(spec) {
  const opts = spec.validate.multipartOptions || {};
  if (typeof opts.autoFields === 'undefined') {
    opts.autoFields = true;
  }
  return async function parseMultipart(ctx, next) {
    if (!ctx.request.is('multipart/*')) {
      return ctx.throw(400, 'expected multipart');
    }
    ctx.request.parts = busboy(ctx, opts);
    await next();
  };
}

/**
 * Creates body parser middleware.
 *
 * @param {Object} spec
 * @return {async function}
 * @api private
 */

function makeBodyParser(spec) {
  if (!(spec.validate && spec.validate.type)) return noopMiddleware;

  switch (spec.validate.type) {
    case 'json':
      return wrapError(spec, makeJSONBodyParser(spec));
    case 'form':
      return wrapError(spec, makeFormBodyParser(spec));
    case 'stream':
    case 'multipart':
      return wrapError(spec, makeMultipartParser(spec));
    default:
      throw new Error(`unsupported body type: ${spec.validate.type}`);
  }
}

/**
 * @api private
 */

function captureError(ctx, type, err) {
  // expose Error message to JSON.stringify()
  err.msg = err.message;
  if (!ctx.invalid) ctx.invalid = {};
  ctx.invalid[type] = err;
}

/**
 * Creates validator middleware.
 *
 * @param {Object} spec
 * @return {async function}
 * @api private
 */

function makeValidator(spec, validatorBuilder) {
  const props = 'header query params body'.split(' ');

  return async function validator(ctx, next) {
    if (!spec.validate) return await next();

    let err;

    for (let i = 0; i < props.length; ++i) {
      const prop = props[i];

      if (spec.validate[prop]) {
        err = validateInput(prop, ctx, spec.validate, validatorBuilder);

        if (err) {
          captureError(ctx, prop, err);
          if (!spec.validate.continueOnError) return ctx.throw(err);
        }
      }
    }

    await next();

    if (spec.validate._outputValidator) {
      debug('validating output');

      err = spec.validate._outputValidator.validate(ctx, validatorBuilder);
      if (err) {
        err.status = 500;
        return ctx.throw(err);
      }
    }
  };
}

/**
 * Exposes route spec
 * @param {Object} spec The route spec
 * @returns {async Function} Middleware
 * @api private
 */
function makeSpecExposer(spec) {
  const defn = clone(spec);
  return async function specExposer(ctx, next) {
    ctx.state.route = defn;
    await next();
  };
}

/**
 * Middleware which creates `request.params`.
 *
 * @api private
 */

async function prepareRequest(ctx, next) {
  ctx.request.params = ctx.params;
  await next();
}

/**
 * Validates request[prop] data with the defined validation schema.
 *
 * @param {String} prop
 * @param {koa.Request} request
 * @param {Object} validate
 * @returns {Error|undefined}
 * @api private
 */

function validateInput(prop, ctx, validate, validatorBuilder) {
  debug('validating %s', prop);

  const request = ctx.request;
  const validator = validatorBuilder(validate[prop]);
  const res = validator(request[prop]);

  if (res.error) {
    res.error.status = validate.failure;
    return res.error;
  }

  // update our request w/ the casted values
  switch (prop) {
    case 'header': // request.header is getter only, cannot set it
    case 'query': // setting request.query directly causes casting back to strings
      Object.keys(res.value).forEach((key) => {
        request[prop][key] = res.value[key];
      });
      break;
    case 'params':
      request.params = ctx.params = res.value;
      break;
    default:
      request[prop] = res.value;
  }
}

/**
 * Routing shortcuts for all HTTP methods
 *
 * Example:
 *
 *    var admin = router();
 *
 *    admin.get('/user', async function(ctx) {
 *      ctx.body = ctx.session.user;
 *    })
 *
 *    var validator = Joi().object().keys({ name: Joi.string() });
 *    var config = { validate: { body: validator }};
 *
 *    admin.post('/user', config, async function(ctx){
 *      console.log(ctx.body);
 *    })
 *
 *    async function commonHandler(ctx){
 *      // ...
 *    }
 *    admin.post('/account', [commonHandler, async function(ctx){
 *      // ...
 *    }]);
 *
 * @param {String} path
 * @param {Object} [config] optional
 * @param {async function|async function[]} handler(s)
 * @return {App} self
 */

methods.forEach((method) => {
  method = method.toLowerCase();

  Router.prototype[method] = function(path) {
    // path, handler1, handler2, ...
    // path, config, handler1
    // path, config, handler1, handler2, ...
    // path, config, [handler1, handler2], handler3, ...

    let fns;
    let config;

    if (typeof arguments[1] === 'function' || Array.isArray(arguments[1])) {
      config = {};
      fns = slice(arguments, 1);
    } else if (typeof arguments[1] === 'object') {
      config = arguments[1];
      fns = slice(arguments, 2);
    }

    const spec = {
      path: path,
      method: method,
      handler: fns
    };

    Object.assign(spec, config);

    this.route(spec);
    return this;
  };
});
